File: src/android/lukuid-sdk/src/test/java/com/lukuid/sdk/GuardCardVocTest.kt
// SPDX-License-Identifier: Apache-2.0
package com.lukuid.sdk

import com.lukuid.sdk.internal.LukuCodec
import com.lukuid.sdk.proto.LukuIDProto
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.security.KeyPairGenerator
import java.security.Signature
import java.util.Base64

class GuardCardVocTest {
    private val magic = byteArrayOf(0x4C, 0x55, 0x4B, 0x55, 0x49, 0x44, 0x01, 0x7E)

    @Test
    fun testDecoderMapsVocRawAndVocIndexFromEnvironmentFrames() {
        val payload = LukuIDProto.EnvironmentPayload.newBuilder()
            .setCtr(4502)
            .setTimestampUtc(1770823456)
            .setVocRaw(30000)
            .setVocIndex(137)
            .build()
        val envRecord = LukuIDProto.EnvironmentRecord.newBuilder()
            .setVersion("1.0.0")
            .setEventId("ENV-VOC-1")
            .setCanonicalString("env-voc-canonical")
            .setPayload(payload)
            .build()
        val response = LukuIDProto.CommandResponse.newBuilder()
            .setStatus(LukuIDProto.Status.STATUS_OK)
            .setSuccess(true)
            .setEnvRecord(envRecord)
            .build()

        var decoded: Map<String, Any?>? = null
        val codec = LukuCodec(
            onMessage = { decoded = it },
            onError = { throw it }
        )
        codec.feed(frame(response.toByteArray()))

        val envMap = decoded?.get("env_record") as Map<*, *>
        val envPayload = envMap["payload"] as Map<*, *>
        assertEquals(30000, envPayload["voc_raw"])
        assertEquals(137, envPayload["voc_index"])
        assertEquals("env-voc-canonical", envMap["canonical_string"])
    }

    @Test
    fun testVerifyEnvelopeAcceptsNewVocCanonicalAndRejectsOldFormat() {
        val keyPair = KeyPairGenerator.getInstance("Ed25519").generateKeyPair()
        val publicKeyRaw = keyPair.public.encoded.copyOfRange(keyPair.public.encoded.size - 32, keyPair.public.encoded.size)
        val publicKeyBase64 = Base64.getEncoder().encodeToString(publicKeyRaw)
        val canonical = "GC-TEST-1:$publicKeyBase64:environment:ENV-VOC-1:4502:1770823456:3600000000:85:false:350.50:22.40:45.20:1013.20:30000:110:false:0.01:0.02:1.00:genesis_fake"
        val signature = signCanonical(keyPair.private.encoded, canonical)

        val envelope = mapOf(
            "type" to "environment",
            "event_id" to "ENV-VOC-1",
            "device_id" to "GC-TEST-1",
            "public_key" to publicKeyBase64,
            "signature" to signature,
            "previous_signature" to "genesis_fake",
            "canonical_string" to canonical,
            "payload" to mapOf(
                "ctr" to 4502,
                "timestamp_utc" to 1770823456,
                "uptime_us" to 3600000000L,
                "battery_percent" to 85,
                "vbus_present" to false,
                "lux" to 350.5,
                "temp_c" to 22.4,
                "humidity_pct" to 45.2,
                "pressure_hpa" to 1013.2,
                "voc_raw" to 30000,
                "voc_index" to 110,
                "tamper" to false,
                "accel_g" to mapOf("x" to 0.01, "y" to 0.02, "z" to 1.0),
                "genesis_hash" to "genesis_fake"
            )
        )

        val validIssues = LukuFile.verifyEnvelope(envelope, LukuVerifyOptions(
            allowUntrustedRoots = true,
            skipCertificateTemporalChecks = true,
            trustProfile = "dev"
        ))
        assertTrue(validIssues.isEmpty())

        val oldCanonical = "GC-TEST-1:$publicKeyBase64:environment:ENV-VOC-1:4502:1770823456:3600000000:85:false:350.50:22.40:45.20:1013.20:110:false:0.01:0.02:1.00:genesis_fake"
        val invalidIssues = LukuFile.verifyEnvelope(
            envelope + mapOf("canonical_string" to oldCanonical),
            LukuVerifyOptions(
                allowUntrustedRoots = true,
                skipCertificateTemporalChecks = true,
                trustProfile = "dev"
            )
        )
        assertTrue(invalidIssues.any { it.code == "RECORD_SIGNATURE_INVALID" })
    }

    private fun frame(payload: ByteArray): ByteArray {
        val out = ByteBuffer.allocate(magic.size + 4 + payload.size + magic.size).order(ByteOrder.LITTLE_ENDIAN)
        out.put(magic)
        out.putInt(payload.size)
        out.put(payload)
        out.put(magic)
        return out.array()
    }

    private fun signCanonical(privateKeyEncoded: ByteArray, canonical: String): String {
        val keyFactory = java.security.KeyFactory.getInstance("Ed25519")
        val privateKey = keyFactory.generatePrivate(java.security.spec.PKCS8EncodedKeySpec(privateKeyEncoded))
        val signer = Signature.getInstance("Ed25519")
        signer.initSign(privateKey)
        signer.update(canonical.toByteArray())
        return Base64.getEncoder().encodeToString(signer.sign())
    }
}

File: src/js/core/test/guardcard_voc.test.ts
// SPDX-License-Identifier: Apache-2.0
import assert from 'node:assert/strict';
import { describe, it } from 'node:test';
import { webcrypto } from 'node:crypto';
import {
  LukuDecoder,
  LukuFile,
  type JsonObject
} from '../src/index.js';

if (!globalThis.crypto) {
  globalThis.crypto = webcrypto as Crypto;
}

const MAGIC = Uint8Array.from([0x4c, 0x55, 0x4b, 0x55, 0x49, 0x44, 0x01, 0x7e]);

function encodeVarint(value: number): number[] {
  const bytes: number[] = [];
  let current = value >>> 0;
  while (current >= 0x80) {
    bytes.push((current & 0x7f) | 0x80);
    current >>>= 7;
  }
  bytes.push(current);
  return bytes;
}

function encodeKey(field: number, wireType: number): number[] {
  return encodeVarint((field << 3) | wireType);
}

function encodeBoolField(field: number, value: boolean): number[] {
  return [...encodeKey(field, 0), ...encodeVarint(value ? 1 : 0)];
}

function encodeUint32Field(field: number, value: number): number[] {
  return [...encodeKey(field, 0), ...encodeVarint(value)];
}

function encodeUint64Field(field: number, value: number): number[] {
  return [...encodeKey(field, 0), ...encodeVarint(value)];
}

function encodeInt64Field(field: number, value: number): number[] {
  return [...encodeKey(field, 0), ...encodeVarint(value)];
}

function encodeStringField(field: number, value: string): number[] {
  const bytes = new TextEncoder().encode(value);
  return [...encodeKey(field, 2), ...encodeVarint(bytes.length), ...bytes];
}

function encodeMessageField(field: number, message: number[]): number[] {
  return [...encodeKey(field, 2), ...encodeVarint(message.length), ...message];
}

function encodeFrame(payload: Uint8Array): Uint8Array {
  const frame = new Uint8Array(MAGIC.length + 4 + payload.length + MAGIC.length);
  let offset = 0;
  frame.set(MAGIC, offset);
  offset += MAGIC.length;
  new DataView(frame.buffer).setUint32(offset, payload.length, true);
  offset += 4;
  frame.set(payload, offset);
  offset += payload.length;
  frame.set(MAGIC, offset);
  return frame;
}

async function createSignedEnvironmentEnvelope(canonicalString: string): Promise<JsonObject> {
  const pair = await crypto.subtle.generateKey({ name: 'Ed25519' }, true, ['sign', 'verify']) as CryptoKeyPair;
  const publicKeyRaw = await crypto.subtle.exportKey('raw', pair.publicKey);
  const publicKeyBase64 = Buffer.from(new Uint8Array(publicKeyRaw)).toString('base64');
  const signature = await crypto.subtle.sign('Ed25519', pair.privateKey, new TextEncoder().encode(canonicalString));

  return {
    type: 'environment',
    id: 'ENV-VOC-1',
    device_id: 'GC-TEST-1',
    public_key: publicKeyBase64,
    signature: Buffer.from(new Uint8Array(signature)).toString('base64'),
    previous_signature: 'genesis_fake',
    canonical_string: canonicalString,
    payload: {
      ctr: 4502,
      timestamp_utc: 1770823456,
      uptime_us: 3600000000,
      battery_percent: 85,
      vbus_present: false,
      lux: 350.5,
      temp_c: 22.4,
      humidity_pct: 45.2,
      pressure_hpa: 1013.2,
      voc_raw: 30000,
      voc_index: 110,
      tamper: false,
      accel_g: { x: 0.01, y: 0.02, z: 1.0 },
      genesis_hash: 'genesis_fake'
    }
  };
}

describe('GuardCard VOC transport', () => {
  it('decodes voc_raw and voc_index from GuardCard environment frames', () => {
    const payload = [
      ...encodeUint64Field(1, 4502),
      ...encodeInt64Field(2, 1770823456),
      ...encodeUint32Field(10, 30000),
      ...encodeUint32Field(21, 137)
    ];
    const envRecord = [
      ...encodeStringField(1, '1.0.0'),
      ...encodeStringField(2, 'ENV-VOC-1'),
      ...encodeStringField(5, 'env-voc-canonical'),
      ...encodeMessageField(6, payload)
    ];
    const response = Uint8Array.from([
      ...encodeVarint((2 << 3) | 0),
      1,
      ...encodeBoolField(3, true),
      ...encodeMessageField(9, envRecord)
    ]);

    const decoder = new LukuDecoder();
    const frames = decoder.push(encodeFrame(response));

    assert.equal(frames.length, 1);
    const envRecordMap = frames[0]?.env_record as JsonObject;
    const envPayload = envRecordMap.payload as JsonObject;
    assert.equal(envPayload.voc_raw, 30000);
    assert.equal(envPayload.voc_index, 137);
    assert.equal(envRecordMap.canonical_string, 'env-voc-canonical');
  });

  it('verifies the new environment canonical string layout and rejects the old one', async () => {
    const unsignedCanonical =
      'GC-TEST-1:PUBLIC_KEY_PLACEHOLDER:environment:ENV-VOC-1:4502:1770823456:3600000000:85:false:350.50:22.40:45.20:1013.20:30000:110:false:0.01:0.02:1.00:genesis_fake';
    const placeholderEnvelope = await createSignedEnvironmentEnvelope(unsignedCanonical);
    const canonical = unsignedCanonical.replace('PUBLIC_KEY_PLACEHOLDER', placeholderEnvelope.public_key as string);
    const envelope = await createSignedEnvironmentEnvelope(canonical);

    const validIssues = await LukuFile.verifyEnvelope(envelope, {
      allowUntrustedRoots: true,
      skipCertificateTemporalChecks: true,
      trustProfile: 'dev'
    });
    assert.equal(validIssues.length, 0, JSON.stringify(validIssues));

    const oldFormatEnvelope: JsonObject = {
      ...envelope,
      canonical_string: `GC-TEST-1:${envelope.public_key as string}:environment:ENV-VOC-1:4502:1770823456:3600000000:85:false:350.50:22.40:45.20:1013.20:110:false:0.01:0.02:1.00:genesis_fake`
    };
    const invalidIssues = await LukuFile.verifyEnvelope(oldFormatEnvelope, {
      allowUntrustedRoots: true,
      skipCertificateTemporalChecks: true,
      trustProfile: 'dev'
    });
    assert.ok(invalidIssues.some((issue) => issue.code === 'RECORD_SIGNATURE_INVALID'));
  });
});
